### PR TITLE
example-message-generator: add support for values(name1 => value1, ...) syntax 

### PR DESCRIPTION
**Pull request by vivinperis**
fixes: https://github.com/syslog-ng/syslog-ng/issues/3215

### CODE REVIEW

**Comment by kira-syslogng** [comment]
This user does not have permission to start the build. Can one of the admins verify this patch and start the build?
(admin: you have the next options (make sure you checked the code):
"ok to test" to accept this pull request (and further changes) for testing
"test this please" for a one time test run
"add to whitelist" add author of a Pull Request to whitelist (globally, be careful, it means this user can trigger kira for any PR)
do nothing -> CI won't start)

**Comment by kira-syslogng** [comment]
This user does not have permission to start the build. Can one of the admins verify this patch and start the build?
(admin: you have the next options (make sure you checked the code):
"ok to test" to accept this pull request (and further changes) for testing
"test this please" for a one time test run
"add to whitelist" add author of a Pull Request to whitelist (globally, be careful, it means this user can trigger kira for any PR)
do nothing -> CI won't start)

**Comment by kira-syslogng** [comment]
This user does not have permission to start the build. Can one of the admins verify this patch and start the build?
(admin: you have the next options (make sure you checked the code):
"ok to test" to accept this pull request (and further changes) for testing
"test this please" for a one time test run
"add to whitelist" add author of a Pull Request to whitelist (globally, be careful, it means this user can trigger kira for any PR)
do nothing -> CI won't start)

**Comment by Kokan** [comment]
@kira-syslogng ok to test

File: modules/examples/sources/msg-generator/msg-generator-source-options.h [hunk 1]
```diff
@@ -79,7 +80,6 @@ msg_generator_source_options_set_template(MsgGeneratorSourceOptions *self, LogTe
 static inline void
 create_log_name_value_pair(MsgGeneratorSourceOptions *self, const gchar *name, const gchar *value)
 {
-  log_msg_set_value_by_name(self->message, name, value, -1);
+  g_hash_table_insert(self->name_value, strdup(name), strdup(value));
```
Comment (Kokan) [inline modules/examples/sources/msg-generator/msg-generator-source-options.h:83]: Please use `g_strdup` (generally we prefer glib provided solutions).

**Comment by kira-syslogng** [comment]
Build SUCCESS

File: modules/examples/sources/msg-generator/msg-generator-grammar.ym [hunk 2]
```diff
@@ -102,10 +103,22 @@ source_msg_generator_option
       msg_generator_source_options_set_template(last_msg_generator_source_options, $3);
       log_template_unref($3);
     }
+  | KW_VALUE '(' options_name_value_pair ')'
   | source_option
   | source_driver_option
   ;
 
+options_name_value_pair
```
Comment (Kokan) [inline modules/examples/sources/msg-generator/msg-generator-grammar.ym:111]: Could you please make the last word plural?

File: modules/examples/sources/msg-generator/msg-generator-grammar.ym [hunk 3]
```diff
@@ -102,10 +103,22 @@ source_msg_generator_option
...
+  : option_name_value_pair options_name_value_pair
+  |
+  ;
+
+option_name_value_pair
+  : string LL_ARROW string_or_number 
+    {      
+      create_log_name_value_pair(last_msg_generator_source_options, $1, $3);
```
Comment (Kokan) [inline modules/examples/sources/msg-generator/msg-generator-grammar.ym:119]: You need to free both `$1` and `$3`.

File: modules/examples/sources/msg-generator/msg-generator-parser.c [hunk 4]
```diff
@@ -35,6 +35,7 @@ static CfgLexerKeyword msg_generator_keywords[] =
   { "freq", KW_FREQ },
   { "num", KW_NUM },
   { "template", KW_TEMPLATE },
+  { "values", KW_VALUE },
```
Comment (Kokan) [inline modules/examples/sources/msg-generator/msg-generator-parser.c:38]: Please us the same name => `KW_VALUES`.

File: modules/examples/sources/msg-generator/msg-generator-source-options.h [hunk 5]
```diff
@@ -26,20 +26,24 @@
...
   LogTemplate *template;
+  GHashTable *name_value;
 } MsgGeneratorSourceOptions;
 
 static inline void
 msg_generator_source_options_defaults(MsgGeneratorSourceOptions *self)
 {
   log_source_options_defaults(&self->super);
   self->freq = 1000;
+  self->name_value = g_hash_table_new(g_str_hash, g_str_equal);
```
Comment (alltilla) [inline modules/examples/sources/msg-generator/msg-generator-source-options.h:46]: Please free the HashTable, when we don't need it anymore.
Also `g_hash_table_new_fullI()` should be used here, so the keys and values are freed as well.

Comment (alltilla) [reply modules/examples/sources/msg-generator/msg-generator-source-options.h:46]: One other thing, I think creating the HashTable should be in `msg_generator_sd_new()`. I know that `msg_generator_source_options_defaults()` is only called there, so the result is the same, but it is an allocation/initialization step, and not a default value setting step.

File: modules/affile/affile-grammar.ym [hunk 6]
```diff
@@ -257,12 +257,15 @@ multi_line_option
...
-	    free($3);
+	    	CHECK_ERROR_GERROR(log_proto_multi_line_server_options_set_garbage(&last_log_proto_options->super, $3, &error), @3, error, "error compiling multi-line regexp");
+	    	free($3);
 	  }
-	| KW_MULTI_LINE_TIMEOUT '(' nonnegative_integer ')'
+	;
+
+multi_line_timeout
+	: KW_MULTI_LINE_TIMEOUT '(' nonnegative_integer ')'
 	  {
```
Comment (alltilla) [inline modules/affile/affile-grammar.ym:267]: Please do not commit these whitespace changes.

File: modules/examples/sources/msg-generator/msg-generator-source-options.h [hunk 7]
```diff
@@ -73,5 +77,9 @@ msg_generator_source_options_set_template(MsgGeneratorSourceOptions *self, LogTe
   log_template_unref(self->template);
   self->template = log_template_ref(template);
 }
-
+static inline void
+create_log_name_value_pair(MsgGeneratorSourceOptions *self, const gchar *name, const gchar *value)
```
Comment (alltilla) [inline modules/examples/sources/msg-generator/msg-generator-source-options.h:81]: I believe this should not be inline, and should be made public in the header file, so `msg-generator-grammar.ym` can reach it.

**Comment by vivinperis** [comment]
I've made the above changes and pushed the new code. Please review it

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by bazsi** [comment]
The branch still contains a copy of your previous fix, please drop that commit from this patchset.

**Pull request closed without merging**